src/crimson/osd/pg_backend.cc

```diff
@@ -80,47 +80,53 @@ PGBackend::load_metadata(const hobject_t& oid)
     coll,
     ghobject_t{oid, ghobject_t::NO_GEN, shard})).safe_then_interruptible(
       [oid](auto &&attrs) -> load_metadata_ertr::future<loaded_object_md_t::ref>{
-	loaded_object_md_t::ref ret(new loaded_object_md_t());
-	if (auto oiiter = attrs.find(OI_ATTR); oiiter != attrs.end()) {
-	  bufferlist bl = std::move(oiiter->second);
-	  ret->os = ObjectState(
-	    object_info_t(bl, oid),
-	    true);
-	} else {
-	  logger().error(
-	    "load_metadata: object {} present but missing object info",
-	    oid);
-	  return crimson::ct_error::object_corrupted::make();
-	}
-	
-	if (oid.is_head()) {
-	  if (auto ssiter = attrs.find(SS_ATTR); ssiter != attrs.end()) {
-	    bufferlist bl = std::move(ssiter->second);
-	    ret->ss = SnapSet(bl);
-	  } else {
-	    /* TODO: add support for writing out snapsets
-	    logger().error(
-	      "load_metadata: object {} present but missing snapset",
-	      oid);
-	    //return crimson::ct_error::object_corrupted::make();
-	    */
-	    ret->ss = SnapSet();
-	  }
-	}
+        loaded_object_md_t::ref ret(new loaded_object_md_t());
+        if (auto oiiter = attrs.find(OI_ATTR); oiiter != attrs.end()) {
+          bufferlist bl = std::move(oiiter->second);
+          ret->os = ObjectState(
+            object_info_t(bl, oid),
+            true);
+        } else {
+          logger().error(
+            "load_metadata: object {} present but missing object info",
+            oid);
+          return crimson::ct_error::object_corrupted::make();
+        }
+
+        if (oid.is_head()) {
+          if (auto ssiter = attrs.find(SS_ATTR); ssiter != attrs.end()) {
+            logger().debug(
+              "load_metadata: object {} and snapset {} present",
+              oid, ssiter->second);
+            bufferlist bl = std::move(ssiter->second);
+            ret->ssc = new crimson::osd::SnapSetContext(oid.get_snapdir());
+            ret->ssc->snapset = SnapSet(bl);
+            if (bl.length()) {
+              ret->ssc->exists = true;
+            } else {
+              ret->ssc->exists = false;
```

Comment:
What does SnapSetContext::exists mean?  IIRC, it's there because a cached SnapSetContext may persist after the logical object to which it belongs is deleted.  If it happens to still be there when the object is read, get_snapset_context will notice and return NULL rather than the context.  For a write, the same context will be returned with exists=false, just as if it were newly created.

PrimaryLogPG::get_snapset_context afaict is using (abusing?) bv.empty() to indicate that the object doesn't exists and can_create.  The bv.length() check should be inaccessible unless either the object exists and bv.length() > 0 or it doesn't exist and can_create in which case bv.length() == 0. (Let me know if I've misread this).

This check is specifically in the path where the object really exists -- the enoent case below handles the object missing case.  Thus, if bl.length() == 0 (or can't be decoded into a SnapSet for any other reason), you should be returning object_corrupted, not an empty context with exists = false.  Also, SnapSet(bl) should be in a try{} block so you can notice a decode error and return an error accordingly.

Reply:
Should I push 7a03a88eca32fac4a29d789ba651d989a63a81a7 as a separate PR as it involves classic osd?